File: connectx-frontend/src/api/userApi.ts
import { apiRequest, uploadRequest } from './apiClient';
import { User } from '../types';

export const userApi = {
  searchUsers: (username: string) =>
    apiRequest<User[]>(`/users/search?username=${encodeURIComponent(username)}`),

  getCurrentUser: () =>
    apiRequest<User>('/users/me'),

  getUserById: (userId: number) =>
    apiRequest<User>(`/users/${userId}`),

  // profileImageUrl is deliberately not accepted here -- profile photos only ever change via
  // uploadProfilePhoto/removeProfilePhoto below, which the backend derives a safe internal path
  // for. Sending an arbitrary profileImageUrl would now be rejected by the backend anyway (the
  // field was removed from UserProfileUpdateDto as part of closing a JWT-exfiltration path).
  updateProfile: (data: { username?: string; displayName?: string; status?: string; profilePhotoVisibility?: string }) =>
    apiRequest<User>('/users/me', {
      method: 'PATCH',
      body: JSON.stringify(data),
    }),

  requestEmailChangeOtp: (newEmail: string) =>
    apiRequest<string>('/users/me/email/request-otp', {
      method: 'POST',
      body: JSON.stringify({ newEmail }),
    }),

  verifyEmailChangeOtp: (newEmail: string, otpCode: string) =>
    apiRequest<User>('/users/me/email/verify-otp', {
      method: 'POST',
      body: JSON.stringify({ newEmail, otpCode }),
    }),

  uploadProfilePhoto: (file: File) => {
    const formData = new FormData();
    formData.append('file', file);
    return uploadRequest<User>('/users/me/profile-photo', formData);
  },

  removeProfilePhoto: () =>
    apiRequest<User>('/users/me/profile-photo', {
      method: 'DELETE',
    }),

  getIdentityKey: () =>
    apiRequest<{ masterPublicKey: string | null; masterPrivateKey: string | null }>('/users/me/identity-key'),

  saveIdentityKey: (data: { masterPublicKey: string; masterPrivateKey: string }) =>
    apiRequest<{ masterPublicKey: string; masterPrivateKey: string }>('/users/me/identity-key', {
      method: 'POST',
      body: JSON.stringify(data),
    }),
};

File: connectx-android/app/src/main/java/com/connectx/app/data/remote/user/UserApi.kt
package com.connectx.app.data.remote.user

import com.connectx.app.core.network.model.ApiResponse
import com.connectx.app.data.remote.auth.model.UserDto
import com.connectx.app.data.remote.user.model.PublicUserDto
import com.connectx.app.data.remote.user.model.RequestEmailChangeOtpRequest
import com.connectx.app.data.remote.user.model.UserIdentityKeyDto
import com.connectx.app.data.remote.user.model.UserProfileUpdateDto
import com.connectx.app.data.remote.user.model.VerifyEmailChangeOtpRequest
import okhttp3.MultipartBody
import okhttp3.ResponseBody
import retrofit2.http.Body
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.Multipart
import retrofit2.http.PATCH
import retrofit2.http.POST
import retrofit2.http.Part
import retrofit2.http.Path
import retrofit2.http.Query
import retrofit2.http.Streaming

/**
 * Full Android-side contract for the backend's User API, as verified directly
 * against com.connectx.user.controller.UserController (10 endpoints) and
 * com.connectx.user.controller.ProfileImageController (1 endpoint) -- 11 total.
 *
 * The N2.0 inspection summary reported "User (10, incl. ProfileImageController)" --
 * a re-count directly from source during N2.2 found UserController alone has 10
 * @*Mapping methods, making 11 total with ProfileImageController. Recorded as a
 * discrepancy in docs/CONNECTX_ANDROID_DEVELOPMENT.md Section 29; this interface
 * reflects the verified source, not the earlier summary.
 *
 * UserDto is reused from data.remote.auth.model -- it is the exact same backend
 * class (com.connectx.user.dto.UserDto) returned by both auth and user endpoints,
 * so it is modeled once, not duplicated.
 *
 * Contract only -- no call site exists anywhere yet. Every one of these 11 endpoints
 * requires authentication on the backend (an authenticated UserPrincipal); no
 * Authorization header handling is added here -- that wiring is deferred to N3.
 */
interface UserApi {

    @GET("api/v1/users/search")
    suspend fun searchUsers(@Query("username") username: String): ApiResponse<List<PublicUserDto>>

    @GET("api/v1/users/me")
    suspend fun getCurrentUser(): ApiResponse<UserDto>

    @GET("api/v1/users/{userId}")
    suspend fun getUserById(@Path("userId") userId: Long): ApiResponse<PublicUserDto>

    @PATCH("api/v1/users/me")
    suspend fun updateProfile(@Body request: UserProfileUpdateDto): ApiResponse<UserDto>

    @POST("api/v1/users/me/email/request-otp")
    suspend fun requestEmailChangeOtp(@Body request: RequestEmailChangeOtpRequest): ApiResponse<String>

    @POST("api/v1/users/me/email/verify-otp")
    suspend fun verifyEmailChangeOtp(@Body request: VerifyEmailChangeOtpRequest): ApiResponse<UserDto>

    @Multipart
    @POST("api/v1/users/me/profile-photo")
    suspend fun uploadProfilePhoto(@Part file: MultipartBody.Part): ApiResponse<UserDto>

    @DELETE("api/v1/users/me/profile-photo")
    suspend fun removeProfilePhoto(): ApiResponse<UserDto>

    @GET("api/v1/users/me/identity-key")
    suspend fun getIdentityKey(): ApiResponse<UserIdentityKeyDto>

    @POST("api/v1/users/me/identity-key")
    suspend fun saveIdentityKey(@Body request: UserIdentityKeyDto): ApiResponse<UserIdentityKeyDto>

    /**
     * Raw binary endpoint (ProfileImageController#getProfileImage) -- returns image
     * bytes directly, never wrapped in ApiResponse. ResponseBody bypasses the
     * registered kotlinx.serialization converter automatically (Retrofit handles
     * ResponseBody natively). @Streaming avoids buffering the whole image in memory.
     */
    @Streaming
    @GET("api/v1/profile-images/{userId}")
    suspend fun getProfileImage(@Path("userId") userId: Long): ResponseBody
}
